### DISCUSSION TITLE
Gdstk vs kfactory

### DISCUSSION THREAD

**Discussion by treep2** [gdsfactory/gdsfactory#2865] [General]
I'll be honest, I have not been up to date on the latest developments of GDSfactory. I was using V6 for a long time which I managed to get to work for many of my personal needs. Yesterday I decided its time for an upgrade and I moved to V8.

Except, none of my codes work! And the entire backend of gdsfactory changed meaning I cannot do my fixes I made in V6, and need to relearn the backend structure.

Why are we suddenly using the kfactory backend? If I want to use kfactory I would use kfactory... 

Current issues I have at the moment, which I have unsuccessfully tried to 'hack' to get back the basic functionality gdsfactory offered:
- Fixing the snapping errors... Why am I not allowed to make a 0.055 nm wide waveguide when I'm using a 1 nm grid? 
- Adding paths over polygons, with the gdstk backend this was a trivial hack (you only needed to add 2 lines of code to the path.py). Paths are a MUST for most ebeam systems utilising FBMS. Without it gdsfactory is practically unusable for most ebeam users.

To be honest, I've spent the past 4 hours fixing the first issue, haven't even had time to look at the second.

I'm moving to GDSfactory V7 tomorrow.

**Comment by tvt173** [comment]
> I'll be honest, I have not been up to date on the latest developments of GDSfactory. I was using V6 for a long time which I managed to get to work for many of my personal needs. Yesterday I decided its time for an upgrade and I moved to V8.
> Except, none of my codes work! And the entire backend of gdsfactory changed meaning I cannot do my fixes I made in V6, and need to relearn the backend structure.

Yes, jumping two major versions at once is a good way to make all your code break :)
Sorry, we put considerable effort in both making the v6->7 and v7->8 transition as smooth as possible, but in the end, this is an open source project that we provide to you for free. There's a limit to what we can reasonably do

> Why are we suddenly using the kfactory backend? If I want to use kfactory I would use kfactory...

It may seem sudden to you, but this is a transition that we spent years debating within the inner circle of maintainers, and there have been various public issues and discussions on the topic over that timespan as well. We made the decision based on our confidence in the core tools and their ability to address all the issues at hand in the long run, as well as managing our own bandwidth and reducing the number of tools we are maintaining. You are of course free to use kfactory or gdsfactory 7 or 6 or 5 or whatever. But what the gdsfactory maintainers have agreed to maintain is gdsfactory 8, and so if you would like to continue to benefit from our efforts, we would recommend you use gdsfactory 8.

> Fixing the snapping errors... Why am I not allowed to make a 0.055 nm wide waveguide when I'm using a 1 nm grid?

a 0.055 nm waveguide on a 1 nm grid? on the surface, this seems insane. what are you trying to accomplish?

> Adding paths over polygons, with the gdstk backend this was a trivial hack (you only needed to add 2 lines of code to the path.py). Paths are a MUST for most ebeam systems utilising FBMS. Without it gdsfactory is practically unusable for most ebeam users.

I'm not the expert on this as I hardly ever use e-beam or paths, but klayout certainly supports paths, so I'm sure it's not impossible here either. @sebastian-goeldi might know better how gds paths can be drawn with kfactory (i don't believe it's exposed in gdsfactory)

> To be honest, I've spent the past 4 hours fixing the first issue, haven't even had time to look at the second.
> I'm moving to GDSfactory V7 tomorrow.

sorry for your difficulty. again, you are welcome to continue to use gdsfactory v7... we even spun out the [gdsfactory7](https://github.com/gdsfactory/gdsfactory7) fork for those who might wish to do so. but our maintenance efforts will be focused on v8

**Reply by treep2** [reply]
> Sorry, we put considerable effort in both making the v6->7 and v7->8 transition as smooth as possible, but in the end, this is an open source project that we provide to you for free. There's a limit to what we can reasonably do

Just pulled the gdsfactory v7, ran it on some of my test codes and so far have not found any issues. The same testcodes I could not get working to my satisfaction on v8.

> It may seem sudden to you, but this is a transition that we spent years debating within the inner circle of maintainers, and there have been various public issues and discussions on the topic over that timespan as well.

Did the inner circle of maintainers unanimously agree on this being the best course of action? Don't get me wrong, I truly appreciate the time and effort many developers are putting into the development of gdsfactory. However I feel this latest update is removing a lot of freedom gdsfactory brought before.

> a 0.055 nm waveguide on a 1 nm grid? on the surface, this seems insane. what are you trying to accomplish?

Sorry, 55 nm waveguide. I don't even care about the 1 nm, the issue is I am using codes to generate certain structures such as tapers which have many points. An example of a multi-segment taper (which I just generated in GDSfactoryV7):
![image](https://github.com/gdsfactory/gdsfactory/assets/172617708/12a3374f-672f-4cd3-9acb-c2b4283dbd85)
This is a headache to make in GDSfactoryV8. I almost got it to work but after all the work decided that I as a user don't want to waste the time in the future to deal with the forcing of symmetry.

My main issues with V8:
Why produce an error when your waveguide is not even? It's possible to make a 55 nm, or 255, or 1005 nm structures in klayout, and if it needs to be snapped to a width of 56 or 54 nm, I don't care. Make this a warning, not an error.
Why produce an error when my waveguide port widths are mismatched by 1 nm? Even the best EUV steppers of ASML (which nobody in photonics will be using for the coming decades) is not able to manufacture a 1 nm offset. I understand these could cause snapping errors when running DRC checks for most fabs. But not everybody cares, just make this a warning. Also I feel making a de-snapping algorithm should be simple, I'll see if I can write this for GDSfactoryV7 when I have time.

I will keep monitoring V8, hopefully the functionality can be improved. For the time being I'm sticking to V7.

**Reply by sebastian-goeldi** [reply]
> Why are we suddenly using the kfactory backend? If I want to use kfactory I would use kfactory...

So why were you not using gdstk before, or to even go further, why were you not just writing machine code? :smile: 
This feels like this age old argument https://xkcd.com/378/
Some of the big reasons for moving towards a klayout powered backend are mentioned in the first paragraph of the readme/docs:

>> This change brings enhanced routing functions and additional features from KLayout, including DRC, dummy fill, and connectivity checks.

> Sorry, 55 nm waveguide. I don't even care about the 1 nm, the issue is I am using codes to generate certain structures such as tapers which have many points. An example of a multi-segment taper (which I just generated in GDSfactoryV7):

There are good reasons for this. gdsfactory7 already started to implement some of these restrictions towards the later versions. Not sure why yours is not flagging, maybe that specific issue was never encountered for the standard.

To show how not so easy snapping (even when you have an integer grid like gds, and becomes magnitues harder when using floating grids like many 3D cad etc) just open the issues page and search for `snapping` in the closed issues. It's not a just "I can write a patch for it" kinda problem. If it were that easy we wouldn't have these restrictions. That or all the contributors of gdsfactory are dummies (possible but unlikely considering the project is now open source for 4+ years and even longer internally, and we to my knowledge we never got a "you dummies why are you not snapping like this" issue). The problem might be well defined for straight waveguides but for the general problem there is no good and easy solution. It all boils down to handling edge cases where you need to round in the correct direction. Sure a human can easily say "oh that's easy just round it that way in this case", but that's not scalable unless you want to check a few thousand to hundreds of thousands of port interfaces for each tapeout.

At my current company we almost lost a very expensive tapeout, all because we had such a case where the rounding did the wrong thing and DRC did in fact not catch it, only an OPC engineer at the fab did.

If you still think the problem is that easy to solve in the general case, kfactory is happily accepting PRs or I am willing to even convert any gdsfactory code for it to kfactory.

If you were to just disable this check in kfactory, you will get this

![image](https://github.com/gdsfactory/gdsfactory/assets/44963764/398c2ccc-40e3-40e5-b021-d6040139cd54)

But then again, in the straight waveguide e.g. you can get away with this, but then the next user  comes and wants off-grid in the other direction (or in a bend etc.), and tada, you have snapping issues

**Comment by treep2** [comment]
Oke, if this fixes snapping issues forever, I'm in. Just not sure how this grid thing will fix snapping issues but I will give it a try.

One question, could one of you add the path functionality? I'm not a contributor, but think this is essential...

The two files which should be changed: 
In components.py add a function for generating paths. I put it after the function 'add_polygon' on line 370:

      def add_path(
          self,
          points: np.ndarray|
          list[list[float]],
          width:float,
          layer: LayerSpec,
      ) -> kdb.Shape:
          """Adds a Polygon to the Component and returns a klayout Shape.
  
          Args:
              points: Coordinates of the vertices of the Polygon.
              layer: layer spec to add polygon on.
          """
          from gdsfactory.pdk import get_layer
  
          layer = get_layer(layer)
  
          if isinstance(points, tuple | list | np.ndarray):
              points = ensure_tuple_of_tuples(points)
              path = kf.kdb.DPath(pts=points,width=width)
  
          return self.shapes(layer).insert(path)



In the path.py file, I change 2 things:

In the function def extrude():
after the if with_simplify  function on line 980:

  if with_simplify:
            points1 = _simplify(points1, tolerance=with_simplify)
            points2 = _simplify(points2, tolerance=with_simplify)

        ## Adding path option in case the width of a section is 0
        Path = False
        if(type(width) == np.ndarray):
            if(width[0] == 0):
                Path=True
        else:
            if(width==0):
                Path=True
              
        ## Either add a path or add a polynomial
        if(Path):
            c.add_path(points1,width=0,layer=layer)
            layers = layer if hidden else [layer,layer]
        else:
            # Join points together
            points_poly = np.concatenate([points1, points2[::-1, :]])

            if not hidden and p_sec.length() > 1e-3:
                c.add_polygon(points_poly, layer=layer)



In the function def extrude_transition():
line 1171 in the function if not hidden and p_sec.length() > 1e-3:

        if not hidden and p_sec.length() > 1e-3:
            if((width1 == 0) | (width2 == 0)): ## Add a path if 0 width
                c.add_path(points_poly, layer=layer,width=0)
            else:
                c.add_polygon(points_poly, layer=layer)

            
It would greatly help me get my colleagues on boards easier. 
Thanks

**Reply by sebastian-goeldi** [reply]
For any manhattan component, I would bet a lot of money, that this fixes any potential snapping issues. I have done now several big tapeouts with this system and can confirm, it's almost impossible to make mistakes, unless you do funny stuff such as defining lengths which aren't mappable to the dbu grid.

As for path, I am not really an expert in which cases they were used in gdsfactory before.

If you are referring to the path as in the geometric object of GDSII, they can be used as follows:

```python
path = gf.kdb.DPath(pts=[gf.kdb.DPoint(x,y) for x,y in my_list_of_tuple_points], width=path_wodth_in_um)
c.shapes(LAYER.WG).insert(path)
```

You can also create it as a list of list of points in the more recent versions:

```python
path = gf.kdb.DPath(pts=my_list_of_tuple_points, width=path_wodth_in_um)
c.shapes(LAYER.WG).insert(path)
```

But this is rather new behavior.

If gds paths are not what you are referring to, @joamatab probably know what it should be

**Reply by treep2** [reply]
Hello Sebastian,

Indeed its these paths I'm referring to, and I'm using this new kfactory DPath function to implement it in gdsfactoryV8. However I need it to be compatible with cross-sections. Examples of thing we try to do:

![image](https://github.com/gdsfactory/gdsfactory/assets/172617708/4e224528-7e4c-4333-af53-3759c628fc4f)

Or if we want to generate a spiral:
I use this code:
`
import gdsfactory as gf
from gdsfactory.cross_section import cross_section


def SiN_FBMS(width: float = 1.0, clad_width: float = 3.0):
    offset = (width + clad_width) / 2
    s1 = gf.Section(width=width + 2 * clad_width, offset=0, layer=(5,0), name='clad')
    s2 = gf.Section(width=0.0, offset=offset, layer=(6,0), name='fbms_L')
    s3 = gf.Section(width=0.0, offset=-offset, layer=(6,0), name='fbms_R')
    return cross_section(width=width, offset=0, layer=(4,0), port_names=['o1', 'o2'], sections=[s1,s2,s3], radius=50.0)

Spiral_c = gf.components.spiral_double(min_bend_radius=100,number_of_loops=3,separation=8,cross_section=SiN_FBMS)

c = gf.Component()
c << Spiral_c
c.show()
`
![image](https://github.com/gdsfactory/gdsfactory/assets/172617708/8877234c-e705-4085-b348-1855f2066848)

**Reply by sebastian-goeldi** [reply]
Hmmm, I see. I am not aware whether this is currently supported or not in gdsfactory8. We are planning on making the general CrossSection more generic and allow more complex constructs to use the basis of a CrossSection (i.e. its sections but without a "core width"). As this would require to limit the generic CrossSection to only sections with non-zero widths.

We would however provide extrude functions for the non-standard CrossSection types such as the asymmetric ones, the ones with components along a path and your case here. I think all of these cases are much closer to something that is exclusively tied to path exctraction and much less tied to path extrusion. (All of this is not settled or even implemented yet though).

cc @joamatab as he can give better input here as well wrt how to create such a path in the current state of gdsfactory

**Reply by treep2** [reply]
Sure lets see what he thinks.

Just to be clear. it's currently not supported in GDSfactoryV8, but I have added it myself (using the few code-lines I've shown in a comment above). 

Indeed, using my implementation doesn't allow for the creation of polygons with a width of 0... But I feel that having paths in gdsfactory will be way more interesting than a polygon with a width of 0.
